examples/hosts/app-integration/external-data/src/utilities/clientManager.ts

```diff
@@ -0,0 +1,122 @@
+ */
+export type ExternalTaskListId = string;
+
+/**
+ * Mock client manager.
+ * Can be registered with to receive notifications when {@link ClientManager.notifySubscribers} is called.
+ *
+ * @typeParam TData - The kind of data that will be sent alongside notifications to subscribers.
+ * Must be JSON-serializable.
+ */
+export class ClientManager<TData = unknown> {
+	/**
+	 * Map of active external resource id to client sessions.
+	 * Values are the set of Fluid Container URLs that will be notified of changes.
+	 */
+	private readonly _taskListMapping: Map<ExternalTaskListId, Set<ClientSessionUrl>>;
+	/**
+	 * Map of active clients to external resource id.
+	 * Values are the set of external resource id's that the client has active and is registered to listen for.
+	 */
+	private readonly _clientMapping: Map<ClientSessionUrl, Set<ExternalTaskListId>>;
+
+	public constructor() {
+		this._clientMapping = new Map<ClientSessionUrl, Set<ExternalTaskListId>>();
+		this._taskListMapping = new Map<ExternalTaskListId, Set<ClientSessionUrl>>();
+	}
+
+	/**
+	 * Gets the current list of client session URLs for the specified task list id.
+	 */
+	public getClientSessions(externalTaskListId: ExternalTaskListId): Set<ClientSessionUrl> {
+		const clientSessionUrls = this._clientMapping.get(externalTaskListId);
+		if (clientSessionUrls === undefined) {
+			console.error(
+				`CUSTOMER SERVICE: "${externalTaskListId}" is not registered to a client session.`,
+			);
+		}
+		return clientSessionUrls ?? new Set<ClientSessionUrl>();
+	}
+
+	/**
+	 * Checks if the client and the external resource already have a mapping and returns the result.
+	 * The user can determine whether to re-register given this information.
+	 */
+	public needsNewMappingEntry(
+		client: ClientSessionUrl,
+		externalTaskListId: ExternalTaskListId,
+	): boolean {
+		let needsNewMappingEntry = true;
+		const clientTaskListIds = this._clientMapping.get(client);
+		// eslint-disable-next-line @typescript-eslint/prefer-optional-chain
+		if (clientTaskListIds !== undefined && clientTaskListIds.has(externalTaskListId)) {
+			const taskListClients = this._taskListMapping.get(client);
+			// eslint-disable-next-line @typescript-eslint/prefer-optional-chain
+			if (taskListClients !== undefined && taskListClients.has(client)) {
+				needsNewMappingEntry = false;
+			}
+		}
+		return needsNewMappingEntry;
+	}
+
+	/**
+	 * Returns a boolean if externalTaskListId already exists entry exists. This means that the customer service
+	 * is already subscribed for webhook notifications for it so we do not need to re-subscribe.
+	 */
+	public needsNewSubscription(externalTaskListId: ExternalTaskListId): boolean {
+		const externalTaskListIdClients = this._taskListMapping.get(externalTaskListId);
+		if (externalTaskListIdClients !== undefined) {
+			return externalTaskListIdClients.size > 0;
+		}
+		return true;
+	}
```

Comment:
Hmm true. Okay so the deletion should then eject the externalTaskListId once there is an empty set after deregistering. Then we can trust !this._taskListMapping.has(externalTaskListId). Let me make sure that's the case below in the deregistering function.

Reply:
I would say don't delete the empty key when the last set entry disappears, because we can't tie that to an actual de-registration with the external data service. So if the code gets to that point, the actual registration in the external data service is still active, but client manager will start reporting that the task list id needs a new subscription. So the fact that the task list id was added to this internal map at some point, tells us that we already called the external service. We should keep it there until we have a way to actually un-register with the external service.